sources/models/src/modeled_types/kubernetes.rs

```diff
@@ -437,3 +437,278 @@ mod test_kubernetes_bootstrap_token {
+            inner: input.to_string(),
+        })
+    }
+}
+string_impls_for!(KubernetesThresholdValue, "KubernetesThresholdValue");
+
+#[cfg(test)]
+mod test_kubernetes_threshold_value {
+    use super::KubernetesThresholdValue;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn good_kubernetes_threshold_value() {
+        for ok in &[ "10%", "10Mi", "1024M", "1Gi" ] {
+            KubernetesThresholdValue::try_from(*ok).unwrap();
+        }
+    }
+
+    #[test]
+    fn bad_kubernetes_threshold_value() {
+        for err in &["", "anything%", ".bad", "bad.", &"a".repeat(64)] {
+            KubernetesThresholdValue::try_from(*err).unwrap_err();
+        }
+    }
+}
+
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// ReservedResourcesKey represents a string that contains a valid Kubernetes kubeReserved
+/// and systemReserved resources i.e. cpu, memory.
+/// https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ReservedResourcesKey {
+    inner: String,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Deserialize)]
+#[serde(rename_all = "lowercase")]
+enum ReservedResources {
+    Cpu,
+    Memory,
+    #[serde(rename = "ephemeral-storage")]
+    EphemeralStorage,
+}
+
+impl TryFrom<&str> for ReservedResourcesKey {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+
+        serde_plain::from_str::<ReservedResources>(&input).context(error::InvalidPlainValue {
+            field: "Reserved sources key",
+        })?;
+        Ok(ReservedResourcesKey {
+            inner: input.to_string(),
+        })
+    }
+}
+string_impls_for!(ReservedResourcesKey, "ReservedResourcesKey");
+
+#[cfg(test)]
+mod test_reserved_resources_key {
+    use super::ReservedResourcesKey;
+    use std::convert::TryFrom;
+
+    #[test]
+    fn good_reserved_resources_key() {
+        for ok in &["cpu", "memory", "ephemeral-storage"] {
+            ReservedResourcesKey::try_from(*ok).unwrap();
+        }
+    }
+
+    #[test]
+    fn bad_reserved_resources_key() {
+        for err in &["", "cpa", ".bad", "bad.", &"a".repeat(64)] {
+            ReservedResourcesKey::try_from(*err).unwrap_err();
+        }
+    }
+}
+/// // =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// KubernetesQuantityValue represents a string that contains a valid kubernetes quantity value.
+/// /// https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct KubernetesQuantityValue {
+    inner: String,
+}
+
+impl TryFrom<&str> for KubernetesQuantityValue {
+    type Error = error::Error;
+
+    fn try_from(input: &str) -> Result<Self, Self::Error> {
+            ensure!(
+                KUBERNETES_QUANTITY.is_match(input),
+                error::InvalidKubernetesQuantity { input }
```

Comment:
Instead of making a new error variant for this, let's use `error::Pattern` pattern here.
```suggestion
            error::Pattern {
                thing: "Kubernetes quantity",
                pattern: KUBERNETES_QUANTITY.clone(),
                input
            }
```